File: apps/api/app/schemas/tracker.py
import uuid
from datetime import datetime
from pydantic import BaseModel
from app.models.job import ApplicationStatus


class ApplicationCreate(BaseModel):
    job_id: uuid.UUID
    status: ApplicationStatus = ApplicationStatus.SAVED


class ApplicationUpdate(BaseModel):
    status: ApplicationStatus


class ApplicationNoteAdd(BaseModel):
    stage: str
    content: str


class ContactUpdate(BaseModel):
    recruiter_name: str | None = None
    recruiter_email: str | None = None


class ApplicationResponse(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID
    status: ApplicationStatus
    applied_at: datetime | None = None
    oa_date: datetime | None = None
    interview_date: datetime | None = None
    offer_date: datetime | None = None
    recruiter_name: str | None = None
    recruiter_email: str | None = None
    notes: dict | None = None
    created_at: datetime

    # Job info
    job_title: str | None = None
    job_company: str | None = None

    model_config = {"from_attributes": True}

File: apps/web/src/lib/api/tracker.ts
import api from './client';

export type AppStatus = 'SAVED' | 'APPLIED' | 'OA' | 'INTERVIEW' | 'OFFER' | 'REJECTED';

export interface ApplicationInfo {
  id: string;
  job_id: string;
  status: AppStatus;
  applied_at: string | null;
  oa_date: string | null;
  interview_date: string | null;
  offer_date: string | null;
  recruiter_name: string | null;
  recruiter_email: string | null;
  notes: Record<string, { content: string; created_at: string }[]> | null;
  created_at: string;
  job_title: string | null;
  job_company: string | null;
}

export async function listApplications(): Promise<ApplicationInfo[]> {
  const { data } = await api.get('/tracker');
  return data;
}

export async function createApplication(jobId: string, status: AppStatus = 'SAVED'): Promise<ApplicationInfo> {
  const { data } = await api.post('/tracker', { job_id: jobId, status });
  return data;
}

export async function updateApplicationStatus(id: string, status: AppStatus): Promise<ApplicationInfo> {
  const { data } = await api.patch(`/tracker/${id}/status`, { status });
  return data;
}

export async function addNote(id: string, stage: string, content: string): Promise<ApplicationInfo> {
  const { data } = await api.post(`/tracker/${id}/notes`, { stage, content });
  return data;
}

export async function deleteApplication(id: string): Promise<void> {
  await api.delete(`/tracker/${id}`);
}
